applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -317,6 +333,81 @@ static int handle_at_slmuart(const char *at_cmd, uint32_t *baudrate)
 	return ret;
 }
 
+void silence_timer_handler(struct k_timer *timer)
+{
+	ARG_UNUSED(timer);
+
+	/* subordinate silence time-out */
+	exit_datamode();
+	if (slm_tcp_get_datamode()) {
+		slm_tcp_set_datamode_off();
+	}
+	if (slm_udp_get_datamode()) {
+		slm_udp_set_datamode_off();
+	}
+	datamode_off_pending = false;
+}
+
+K_TIMER_DEFINE(silence_timer, silence_timer_handler, NULL);
+
+static void raw_send(struct k_work *work)
+{
+	const char *terminator = CONFIG_SLM_DATAMODE_TERMINATOR;
+	int len = strlen(terminator);
+	int64_t silence_limit = CONFIG_SLM_DATAMODE_SILENCE * MSEC_PER_SEC;
+	int64_t silence_delta = k_uptime_delta(&silence_start);
+	int err;
+
+	ARG_UNUSED(work);
+
+	LOG_HEXDUMP_DBG(at_buf, at_buf_len, "RX");
+
+	/* check leading silence */
+	if (silence_delta < silence_limit) {
+		if (datamode_off_pending) {
+			uint8_t buf[UART_RX_LEN + len + 1];
+			/* cancel pending datamode off */
+			k_timer_stop(&silence_timer);
+			datamode_off_pending = false;
+			/* insert termiantor string */
+			strcpy(buf, terminator);
+			strcat(buf, at_buf);
+			strcpy(at_buf, buf);
+			at_buf_len += len;
+		}
+	} else {
+		if (at_buf_len == len &&
+		    strncmp(at_buf, terminator, len) == 0) {
```

Comment:
Isn't this oversimplified approach? I see 2 possible risks here:
 * What if the terminator string is not present at the beginning of the buffer?
 * What if the terminator string is split between two UART IRQs?

Reply:
Here we must guarantee received string and terminator are exactly the same, except that the terminator has "\0" when compiler put the CONFIG string in global data area. The terminator string is const and defined as below:
   const char *terminator = CONFIG_SLM_DATAMODE_TERMINATOR;